[Backend] Schema.sql Dollar-Quote Block Crashes Backend In Docker Local Compose

# 📝 Schema.sql Dollar-Quote Block Crashes Backend In Docker Local Compose

## Description
Running `docker compose -f docker-compose.local.yml up --build -d` fails to start the `mapcess_backend_dev` container. The error is thrown from `dataSourceScriptDatabaseInitializer` while loading `backend/src/main/resources/schema.sql`:

```
Caused by: org.postgresql.util.PSQLException: Unterminated dollar quote started at position 3 in SQL DO $$ BEGIN IF EXISTS ( SELECT 1 FROM information_schema.columns WHERE table_name = 'reports' AND column_name = 'publish_date' AND data_type = 'timestamp without time zone' ) THEN ALTER TABLE reports ALTER COLUMN publish_date TYPE TIMESTAMPTZ USING publish_date AT TIME ZONE 'UTC'. Expected terminating $$
```

Spring's `ScriptUtils` (used by `DataSourceScriptDatabaseInitializer`) splits SQL scripts on `;` and only recognizes `'...'` and `"..."` quoting — it has no understanding of PostgreSQL dollar-quoting (`$$ ... $$`). The migration in `schema.sql` (introduced in #394, commit `813d9c2`) wraps multiple `ALTER TABLE` statements inside a `DO $$ BEGIN ... END $$;` block. Spring slices the file at the first `;` it encounters inside the block (after `'UTC';`), and the resulting fragment fails to parse because the `$$` is unclosed. The migration logic itself is valid PostgreSQL — only the Spring-side script loader can't handle it.

The bug ships because the only execution path that triggers it is `docker compose -f docker-compose.local.yml up`, and that path isn't covered by CI or by the most common dev workflows: `backend-pr-check.yml` runs `./mvnw verify` against H2 with `spring.sql.init.mode=never` (set in `application-test.properties`); `application.properties` defaults to `never` so `./mvnw spring-boot:run` doesn't trigger it; only the local compose explicitly sets `SPRING_SQL_INIT_MODE=always`.

## Feature / Task Details
Three workable fixes, ranked by invasiveness:

- **Option 1 — Drop the `DO $$ ... $$;` block from `schema.sql`.** Acceptable because (a) prod was migrated when #394 landed, (b) fresh local DBs get `TIMESTAMPTZ` from Hibernate's `ddl-auto=update` since `Report.publishDate` and `Comment.createdAt` are `Instant`. Lowest-risk; loses the migration safety net for devs with stale local DBs (mitigation: `docker compose down -v` resets).
- **Option 2 — Configure `spring.sql.init.separator=^---^`** (or any unique sentinel) and update both `schema.sql` and `data.sql` to use it as the statement terminator. Preserves the migration; touches ~30 statement terminators in `data.sql`.
- **Option 3 — Move the migration to a Java `@Bean ApplicationRunner`** that opens a JDBC `Statement` and `execute()`s the DO block as one string. Bypasses Spring's splitter entirely. Heaviest but most future-proof; opens the door to Flyway/Liquibase later.

Recommendation: ship Option 1 to unblock local testing now; file a follow-up to introduce Flyway if column-type migrations become a recurring pattern.

## Deliverables
- A `schema.sql` that does not crash at startup with `SPRING_SQL_INIT_MODE=always`.
- If keeping the migration: working timestamp migration for legacy DBs, or a documented manual `psql` script if the migration is removed.
- Backend health verified end-to-end via `docker compose -f docker-compose.local.yml up --build -d`.

## Acceptance Criteria

- [ ] `docker compose -f docker-compose.local.yml up --build -d` starts the backend cleanly with no `Unterminated dollar quote` error.
- [ ] Backend `GET /api/reports` returns `200` after startup.
- [ ] `data.sql` seed still loads correctly (Adminer shows seeded users + reports).
- [ ] All tests pass.
- [ ] Documentation updated (if applicable).

## Deadline
2026-05-10

## Notes
- Refs: #221 (introduced the timestamp-type change), #394 (PR that merged the broken `schema.sql`).
- Discovered while testing #401 (web fix-flow UI) locally — first time anyone re-spun the local compose after #394 merged.

--- a/backend/src/main/resources/data.sql
+++ b/backend/src/main/resources/data.sql
@@ -66,28 +66,28 @@ SELECT setval('report_categories_id_seq', (SELECT MAX(id) FROM report_categories
 -- referenced parent report cannot be found.
 -- Comments: guard with NOT EXISTS on content + report_id
 INSERT INTO comments (content, author_id, report_id, created_at)
-SELECT 'Yes, people also fall from this ramp', 2, r.id, NOW()
+SELECT 'Yes, people also fall from this ramp', 2, r.report_id, NOW()
 FROM reports r
 WHERE r.description = 'This ramp is too steep for wheelchairs'
   AND NOT EXISTS (
     SELECT 1 FROM comments c
-    WHERE c.content = 'Yes, people also fall from this ramp' AND c.report_id = r.id
+    WHERE c.content = 'Yes, people also fall from this ramp' AND c.report_id = r.report_id
   );
 
 INSERT INTO comments (content, author_id, report_id, created_at)
-SELECT 'Please fix this ASAP, it affects many students.', 3, r.id, NOW()
+SELECT 'Please fix this ASAP, it affects many students.', 3, r.report_id, NOW()
 FROM reports r
 WHERE r.description = 'This ramp is too steep for wheelchairs'
   AND NOT EXISTS (
     SELECT 1 FROM comments c
-    WHERE c.content = 'Please fix this ASAP, it affects many students.' AND c.report_id = r.id
+    WHERE c.content = 'Please fix this ASAP, it affects many students.' AND c.report_id = r.report_id
   );
 
 INSERT INTO comments (content, author_id, report_id, created_at)
-SELECT 'The elevator was repaired according to staff.', 3, r.id, NOW()
+SELECT 'The elevator was repaired according to staff.', 3, r.report_id, NOW()
 FROM reports r
 WHERE r.description = 'Elevator of Boğaziçi Metro is broken'
   AND NOT EXISTS (
     SELECT 1 FROM comments c
-    WHERE c.content = 'The elevator was repaired according to staff.' AND c.report_id = r.id
+    WHERE c.content = 'The elevator was repaired according to staff.' AND c.report_id = r.report_id
   );

--- a/backend/src/main/resources/schema.sql
+++ b/backend/src/main/resources/schema.sql
@@ -1,32 +1,3 @@
--- Migrate legacy timestamp columns to TIMESTAMPTZ so the API emits Z-suffixed
--- timestamps via Jackson's Instant serializer. See issue #221.
---
--- Runs after Hibernate (spring.jpa.defer-datasource-initialization=true) and
--- before data.sql. Idempotent: a no-op once the column is already TIMESTAMPTZ.
--- Existing values are reinterpreted as UTC, which matches the production JVM
--- timezone (containers run in UTC) so wall-clock instants are preserved.
-
-DO $$
-BEGIN
-    IF EXISTS (
-        SELECT 1 FROM information_schema.columns
-        WHERE table_name = 'reports'
-          AND column_name = 'publish_date'
-          AND data_type = 'timestamp without time zone'
-    ) THEN
-        ALTER TABLE reports
-            ALTER COLUMN publish_date TYPE TIMESTAMPTZ
-            USING publish_date AT TIME ZONE 'UTC';
-    END IF;
-
-    IF EXISTS (
-        SELECT 1 FROM information_schema.columns
-        WHERE table_name = 'comments'
-          AND column_name = 'created_at'
-          AND data_type = 'timestamp without time zone'
-    ) THEN
-        ALTER TABLE comments
-            ALTER COLUMN created_at TYPE TIMESTAMPTZ
-            USING created_at AT TIME ZONE 'UTC';
-    END IF;
-END $$;
+-- Intentionally empty. Spring's ScriptUtils can't parse PostgreSQL
+-- dollar-quoted blocks (see #404). Hibernate creates TIMESTAMPTZ
+-- columns directly from Instant entity fields, so no migration runs here.

--- a/backend/src/main/resources/schema.sql
+++ b/backend/src/main/resources/schema.sql
@@ -1,3 +1,6 @@
--- Intentionally empty. Spring's ScriptUtils can't parse PostgreSQL
--- dollar-quoted blocks (see #404). Hibernate creates TIMESTAMPTZ
--- columns directly from Instant entity fields, so no migration runs here.
+-- Intentionally minimal. Spring's ScriptUtils can't parse PostgreSQL
+-- dollar-quoted blocks (see #404), and rejects comment-only files with
+-- "script must not be null or empty" -- so we keep one no-op statement.
+-- Hibernate creates TIMESTAMPTZ columns directly from Instant entity
+-- fields, so no real migration runs here.
+SELECT 1;

